File: web/src/lib/stores/sabbath.ts
// Sabbath mode — a per-device client-side overlay on the nav + dashboard
// that hides work-oriented modules and surfaces rest-oriented ones.
// Mark 2:27: "the sabbath was made for man."
//
// State splits cleanly in two:
//
//   • Schedule (recurring rule) — synced server-side via /api/v1/sabbath.
//     Changes on one device show up on every other device. This is what
//     "every Saturday from midnight" means; also where sundown-to-sundown
//     windows live.
//
//   • Daily on/off (manual "begin now") — per-device, localStorage only.
//     A user toggling sabbath on their laptop doesn't force it on their
//     phone. The server gates work-modules (push, AI, agents) based on
//     the SCHEDULE alone — if you want gating during a manual sabbath,
//     configure the schedule.
//
// The schedule window can span midnight (Friday 18:00 + 24h goes through
// Saturday 18:00), so traditional sundown-to-sundown observance is a
// first-class shape, not a special case.
//
// Auto-expiry: the read-time `isActiveNow()` check uses (a) ActiveOn ==
// today for the manual flag, and (b) `now ∈ [windowStart, windowEnd)`
// for the schedule. Both fall off naturally when their condition stops
// holding — no background timer required, no leak risk.

import { writable, get, type Readable } from 'svelte/store';
import { loadStoredString, saveStoredString } from '$lib/util/storage';
import { api, type SabbathSchedulePayload } from '$lib/api';

const KEY = 'granit.sabbath.activeOn';
const SCHEDULE_KEY = 'granit.sabbath.schedule';
const SKIP_KEY = 'granit.sabbath.skipOn';

// Modules considered "work" — hidden when sabbath is active. The
// list is intentional, not user-configurable: the point of sabbath
// is *the discipline of letting go*, not picking and choosing.
export const SABBATH_HIDE_MODULES = [
  'finance',
  'tasks',
  'projects',
  'agents',
  'deadlines',
  'chat',
  'weekly_review',
  'emails',
  'shopping',
  'ventures',
  'goals',
  'habits',
  'measurements',
  'objects',
  'hub'
];

// Dashboard widget types hidden during sabbath. Mirror of
// SABBATH_HIDE_MODULES at widget granularity — the modules list
// gates the nav and route guards, this list gates the home-page
// widget grid so a "tasks paused" sabbath doesn't still surface a
// today-tasks widget right under the greeting. The string literals
// match the keys in $lib/dashboard/registry.ts.
//
// today-stream is hidden because it merges events + tasks +
// deadlines into one feed and there's no per-row sabbath filter
// inside it. Surface events via /calendar (route guard exempts
// it — calendar isn't in SABBATH_HIDE_MODULES) if you need them.
// today-focus is hidden because the morning routine commitment is
// a work-day artifact.
export const SABBATH_HIDE_WIDGET_TYPES: ReadonlySet<string> = new Set([
  'today-tasks',
  'scheduled-today',
  'inbox',
  'goals-progress',
  'projects-active',
  'ventures',
  'habits',
  'streaks',
  'weekly-plan',
  'today-focus',
  'today-stream'
]);

// Rest modules surfaced as a hint when sabbath starts — nav doesn't
// hide anything from this list even if the user has them disabled
// in normal config.
export const SABBATH_SURFACE_MODULES = [
  'scripture',
  'prayer',
  'people',
  'vision',
  'jots'
];

export const DAY_LABELS = ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday'];

function todayISO(): string {
  const d = new Date();
  return `${d.getFullYear()}-${String(d.getMonth() + 1).padStart(2, '0')}-${String(d.getDate()).padStart(2, '0')}`;
}

// ── Schedule ────────────────────────────────────────────────────
// Camel-cased TS-side shape. Wire shape (snake_case) lives in
// api.ts as SabbathSchedulePayload; the two converters below adapt
// at the network boundary.
export interface SabbathSchedule {
  enabled: boolean;
  dayOfWeek: number;       // 0=Sun … 6=Sat
  startHour: number;       // 0-23
  startMinute: number;     // 0-59
  durationMinutes: number; // 1440 = 24h (midnight-to-midnight)
}

export const DEFAULT_SCHEDULE: SabbathSchedule = {
  enabled: false,
  dayOfWeek: 0,
  startHour: 0,
  startMinute: 0,
  durationMinutes: 1440
};

function toWire(s: SabbathSchedule): SabbathSchedulePayload {
  return {
    enabled: s.enabled,
    day_of_week: s.dayOfWeek,
    start_hour: s.startHour,
    start_minute: s.startMinute,
    duration_minutes: s.durationMinutes
  };
}

function fromWire(p: SabbathSchedulePayload): SabbathSchedule {
  return {
    enabled: Boolean(p.enabled),
    dayOfWeek: clampInt(p.day_of_week, 0, 6, 0),
    startHour: clampInt(p.start_hour, 0, 23, 0),
    startMinute: clampInt(p.start_minute, 0, 59, 0),
    durationMinutes: clampInt(p.duration_minutes, 1, 7 * 1440, 1440)
  };
}

function clampInt(v: unknown, min: number, max: number, fallback: number): number {
  const n = Number(v);
  if (!Number.isFinite(n)) return fallback;
  const i = Math.round(n);
  if (i < min) return min;
  if (i > max) return max;
  return i;
}

function loadScheduleLocal(): SabbathSchedule {
  const raw = loadStoredString(SCHEDULE_KEY, '');
  if (!raw) return { ...DEFAULT_SCHEDULE };
  try {
    const parsed = JSON.parse(raw);
    return {
      enabled: Boolean(parsed.enabled),
      dayOfWeek: clampInt(parsed.dayOfWeek, 0, 6, 0),
      startHour: clampInt(parsed.startHour, 0, 23, 0),
      startMinute: clampInt(parsed.startMinute, 0, 59, 0),
      durationMinutes: clampInt(parsed.durationMinutes, 1, 7 * 1440, 1440)
    };
  } catch {
    return { ...DEFAULT_SCHEDULE };
  }
}

function persistScheduleLocal(sched: SabbathSchedule) {
  try {
    saveStoredString(SCHEDULE_KEY, JSON.stringify(sched));
  } catch {
    // localStorage full / private mode — non-fatal; the server is the
    // source of truth and will be re-fetched on next load anyway.
  }
}

// ── Schedule window math ────────────────────────────────────────
// Mirror of the Go-side scheduleWindow in internal/sabbath. Finds
// the most-recently-started occurrence of (DayOfWeek, StartHour,
// StartMinute) at or before `at`, returns its [start, end) window.
// Returns null when schedule is disabled or no occurrence fits.
//
// Exported so the midnight-rollover + sundown-to-sundown invariants
// can be unit-tested; consumers outside this module should still
// prefer the `sabbath` store for "is sabbath active right now".
export function scheduleWindow(s: SabbathSchedule, at: Date): { start: Date; end: Date } | null {
  if (!s.enabled || s.durationMinutes <= 0) return null;
  // Walk back up to 8 days so today + every preceding day are covered.
  for (let daysBack = 0; daysBack < 8; daysBack++) {
    const cand = new Date(at.getFullYear(), at.getMonth(), at.getDate() - daysBack,
      s.startHour, s.startMinute, 0, 0);
    if (cand.getDay() !== s.dayOfWeek) continue;
    if (cand.getTime() > at.getTime()) continue;
    const end = new Date(cand.getTime() + s.durationMinutes * 60_000);
    return { start: cand, end };
  }
  return null;
}

export function scheduleSaysNow(s: SabbathSchedule, at: Date): boolean {
  const w = scheduleWindow(s, at);
  if (!w) return false;
  return at.getTime() >= w.start.getTime() && at.getTime() < w.end.getTime();
}

// ── Stores ──────────────────────────────────────────────────────
export const sabbathSchedule = writable<SabbathSchedule>(loadScheduleLocal());

// Persist locally on every change; sync to server when the change
// originates from user action (set via updateSchedule()). The store
// subscribe runs for both kinds of update, so we use a sync-suppression
// flag to avoid PUT-loop on server-driven updates.
let suppressServerSync = false;
sabbathSchedule.subscribe((s) => {
  persistScheduleLocal(s);
  if (suppressServerSync) return;
  // Best-effort PUT. Failure is non-fatal — the local copy is now
  // ahead of the server; next successful PUT (or page load if we
  // adopt last-write-wins later) reconciles.
  api.putSabbath({ schedule: toWire(s) }).catch(() => undefined);
});

// updateSchedule is the user-facing setter. Identical to .set() at
// this layer but reads as intent in call sites.
export function updateSchedule(next: SabbathSchedule) {
  sabbathSchedule.set(next);
}

// ── Daily on/off (per-device manual flag) ───────────────────────
// Local-only. The server gates work-modules based on the synced
// schedule, so a manual toggle here doesn't propagate. If you want
// server gating during a manual sabbath, configure the schedule.
function loadManualActive(): boolean {
  return loadStoredString(KEY, '') === todayISO();
}

// Skip — the "exit sabbath" override. When the user dismisses an
// otherwise-scheduled sabbath, we record today's date here AND PUT
// it to the server so server-side gates (push, agents, AI) also
// stop silencing for the day. The schedule itself is untouched;
// next week's sabbath fires normally.
function loadSkipToday(): boolean {
  return loadStoredString(SKIP_KEY, '') === todayISO();
}

function computeInitial(): boolean {
  if (loadSkipToday()) return false;
  if (loadManualActive()) return true;
  return scheduleSaysNow(loadScheduleLocal(), new Date());
}

const { subscribe, set } = writable<boolean>(computeInitial());

export const sabbath: Readable<boolean> & {
  enable(): void;
  disable(): void;
  toggle(): void;
  isActive(): boolean;
} = {
  subscribe,
  enable() {
    // Re-entering manually also clears the skip — if you change your
    // mind and want sabbath back on today, the skip shouldn't keep
    // suppressing the schedule on the next focus.
    saveStoredString(SKIP_KEY, undefined);
    saveStoredString(KEY, todayISO());
    set(true);
    // Tell the server too so server gates re-engage.
    api.putSabbath({ skip_on: '' }).catch(() => undefined);
  },
  disable() {
    // Clear the manual flag AND record a skip-today. The skip is the
    // escape hatch: without it, an active schedule would re-assert
    // immediately and the user would think the button was broken.
    // PUT the skip to the server so push/agent/AI gates also stop
    // silencing for the day. Schedule itself is untouched — next
    // week's sabbath still fires.
    saveStoredString(KEY, undefined);
    const today = todayISO();
    saveStoredString(SKIP_KEY, today);
    set(false);
    api.putSabbath({ skip_on: today, active_on: '' }).catch(() => undefined);
  },
  toggle() {
    if (get({ subscribe })) this.disable();
    else this.enable();
  },
  isActive() {
    return get({ subscribe });
  }
};

// ── Time-remaining ──────────────────────────────────────────────
// Returns minutes until the current sabbath window ends. For
// manual-only activations falls back to midnight tomorrow (the
// original behavior). Returns 0 when not active.
export function sabbathMinutesRemaining(): number {
  if (!get({ subscribe })) return 0;
  const now = new Date();
  const w = scheduleWindow(get(sabbathSchedule), now);
  if (w && now.getTime() >= w.start.getTime() && now.getTime() < w.end.getTime()) {
    return Math.max(0, Math.round((w.end.getTime() - now.getTime()) / 60_000));
  }
  const tomorrow = new Date(now.getFullYear(), now.getMonth(), now.getDate() + 1, 0, 0, 0, 0);
  return Math.max(0, Math.round((tomorrow.getTime() - now.getTime()) / 60_000));
}

// ── Server sync ─────────────────────────────────────────────────
// Fetch the canonical schedule from the server on first load + on
// every tab-focus. The server is the source of truth for the
// schedule; the local copy is a cache that only matters until the
// fetch resolves.
//
// Conflict resolution: server wins on read. If the user just edited
// the schedule locally and the PUT raced with the next focus-driven
// GET, the local edit might briefly flicker back to the server's
// previous value before the PUT-response reconciles. Acceptable
// tradeoff — the alternative (timestamps + last-write-wins) is more
// machinery than this surface needs.
//
// Coalesce: refetch is fired from the module init AND from every
// visibilitychange. A user toggling between tabs repeatedly while the
// server is slow would otherwise queue overlapping GETs — they all
// land on the same endpoint and the last one's `.set()` wins. The
// inFlight flag drops anything that comes in while a fetch is open;
// the visibilitychange that fires after it lands will pick up any new
// state.
let refetchInFlight = false;
async function refetchSchedule() {
  if (refetchInFlight) return;
  refetchInFlight = true;
  try {
    const res = await api.getSabbath();
    const remote = fromWire(res.schedule);
    const current = get(sabbathSchedule);
    if (!schedulesEqual(remote, current)) {
      suppressServerSync = true;
      try {
        sabbathSchedule.set(remote);
      } finally {
        suppressServerSync = false;
      }
      // Recompute the active flag now that the schedule may have changed.
      set(computeActiveAt(remote, new Date()));
    }
    // Server's skip_on is authoritative — pull it down so other
    // devices honor a skip that originated here, and so a skip set
    // by today's PUT survives a manual page reload.
    if (typeof res.skip_on === 'string') {
      if (res.skip_on) saveStoredString(SKIP_KEY, res.skip_on);
      else if (loadStoredString(SKIP_KEY, '') !== '') saveStoredString(SKIP_KEY, undefined);
    }
  } catch {
    // Server unreachable / auth not yet hydrated — try again on next
    // visibilitychange. The local copy is fine until then.
  } finally {
    refetchInFlight = false;
  }
}

// computeActiveAt is the single source of truth for "is sabbath
// active right now": skip wins → manual wins → schedule. Matches the
// Go IsActiveAt ordering.
function computeActiveAt(s: SabbathSchedule, at: Date): boolean {
  if (loadSkipToday()) return false;
  if (loadManualActive()) return true;
  return scheduleSaysNow(s, at);
}

function schedulesEqual(a: SabbathSchedule, b: SabbathSchedule): boolean {
  return a.enabled === b.enabled
    && a.dayOfWeek === b.dayOfWeek
    && a.startHour === b.startHour
    && a.startMinute === b.startMinute
    && a.durationMinutes === b.durationMinutes;
}

if (typeof window !== 'undefined') {
  // Initial fetch — fire-and-forget, runs as soon as the module loads.
  // The auth token may not be hydrated yet on cold-start; the
  // visibilitychange handler will retry shortly.
  refetchSchedule();

  document.addEventListener('visibilitychange', () => {
    if (document.visibilityState !== 'visible') return;
    // Re-fetch the schedule (server may have changed) and re-evaluate
    // local active state (date may have rolled over, or window may
    // have just opened/closed, or a skip may have expired).
    refetchSchedule();
    const fresh = computeActiveAt(get(sabbathSchedule), new Date());
    if (fresh !== get({ subscribe })) set(fresh);
  });
}

File: internal/sabbath/sabbath.go
// Package sabbath persists the user's Sabbath-mode state to a
// vault-side JSON sidecar so server-side surfaces (push scheduler,
// background agents, AI handlers) can respect it. The frontend drives
// its own UI overlay from localStorage; this package mirrors the
// schedule + the daily flag to disk so the server gates work for the
// user's day of rest regardless of which device is open.
//
// Two pieces of state:
//
//   - ActiveOn (YYYY-MM-DD) — the manual "begin now" flag. Per-device
//     on the UI side, but mirrored here so server-side handlers gate
//     when ANY device is in sabbath. Auto-clears when the date no
//     longer matches today.
//
//   - Schedule {Enabled, DayOfWeek, StartHour, StartMinute,
//     DurationMinutes} — the recurring rule. Synced across devices
//     so a sabbath configured on one phone shows up on the laptop.
//     The window can span midnight (Friday 18:00 + 24h goes through
//     Saturday 18:00) so traditional sundown-to-sundown observance is
//     a first-class shape, not a special case.
//
// IsActiveNow returns true when either: (a) the manual flag matches
// today, or (b) the schedule window covers the current instant. Used
// by every server-side gate (push, AI, chat, agent runs).
//
// Stored at <vault>/.granit/sabbath.json with 0o600 perms via
// atomicio.
package sabbath

import (
	"encoding/json"
	"errors"
	"os"
	"path/filepath"
	"time"

	"github.com/artaeon/granit/internal/atomicio"
)

// Schedule is the recurring auto-enable rule. DayOfWeek matches
// time.Weekday (0=Sunday … 6=Saturday). The window starts at
// StartHour:StartMinute on that day and runs DurationMinutes.
// Default 1440 minutes = 24 hours = traditional midnight-to-midnight
// when StartHour/StartMinute are both 0. Setting StartHour=18 with
// a Friday DayOfWeek and 1440 duration produces a sundown-to-sundown
// Jewish sabbath observance.
//
// Enabled=false means schedule never fires (manual flag is the only
// way into sabbath mode).
type Schedule struct {
	Enabled         bool `json:"enabled"`
	DayOfWeek       int  `json:"day_of_week"`
	StartHour       int  `json:"start_hour"`
	StartMinute     int  `json:"start_minute"`
	DurationMinutes int  `json:"duration_minutes"`
}

// DefaultSchedule is what an unconfigured State returns. Off, but with
// sane field values so a UI binding to the struct doesn't see zeros
// that look like "configured midnight Sunday".
func DefaultSchedule() Schedule {
	return Schedule{Enabled: false, DayOfWeek: 0, StartHour: 0, StartMinute: 0, DurationMinutes: 1440}
}

// State is the persisted Sabbath flag. ActiveOn is the YYYY-MM-DD
// date the user manually enabled the mode (empty → manual flag off).
// Schedule is the recurring auto-enable rule.
type State struct {
	ActiveOn string   `json:"active_on,omitempty"`
	Schedule Schedule `json:"schedule"`
}

// Path returns the absolute path of the sidecar.
func Path(vaultRoot string) string {
	return filepath.Join(vaultRoot, ".granit", "sabbath.json")
}

// Load reads the sidecar. Missing file → State with DefaultSchedule
// (not an error). Missing schedule field → DefaultSchedule applied so
// older sidecars (pre-schedule) read cleanly without a migration.
func Load(vaultRoot string) (State, error) {
	data, err := os.ReadFile(Path(vaultRoot))
	if err != nil {
		if errors.Is(err, os.ErrNotExist) {
			return State{Schedule: DefaultSchedule()}, nil
		}
		return State{}, err
	}
	var s State
	if err := json.Unmarshal(data, &s); err != nil {
		return State{}, err
	}
	// Pre-schedule sidecars deserialize with a zero Schedule. The
	// zero value has DurationMinutes=0 which would make IsActiveNow
	// reject every instant — distinguishable from the explicit
	// "schedule off" intent by treating a zero-duration as a missing
	// field and substituting the default.
	if s.Schedule.DurationMinutes == 0 {
		s.Schedule = DefaultSchedule()
		s.Schedule.Enabled = false
	}
	return s, nil
}

// Save persists the state.
func Save(vaultRoot string, s State) error {
	dir := filepath.Dir(Path(vaultRoot))
	if err := os.MkdirAll(dir, 0o700); err != nil {
		return err
	}
	data, err := json.MarshalIndent(s, "", "  ")
	if err != nil {
		return err
	}
	return atomicio.WriteState(Path(vaultRoot), data)
}

// scheduleWindow returns the [start, end) window of the schedule's
// CURRENT or MOST-RECENTLY-STARTED occurrence relative to `at`. The
// idea: find the most recent schedule-day occurrence whose start is
// in the past, and return its window. If `at` falls inside that
// window we're in sabbath. If not, we aren't.
//
// Why most-recent-past instead of "today's instance": a Friday 18:00
// → Saturday 18:00 window straddles Friday and Saturday calendar
// days. At Saturday 10:00 the "today's start" would be Saturday
// 00:00 which is wrong; we need to look back at Friday's start.
//
// Returns ok=false when schedule is disabled.
func scheduleWindow(s Schedule, at time.Time) (start, end time.Time, ok bool) {
	if !s.Enabled || s.DurationMinutes <= 0 {
		return time.Time{}, time.Time{}, false
	}
	// Find the most recent occurrence of (DayOfWeek, StartHour,
	// StartMinute) at or before `at`. Walk back at most 8 days from
	// today to cover the case where today's schedule-time hasn't hit
	// yet — we'd then use last week's window (already past, won't
	// match) but the function stays correct.
	for daysBack := 0; daysBack < 8; daysBack++ {
		cand := time.Date(at.Year(), at.Month(), at.Day()-daysBack,
			s.StartHour, s.StartMinute, 0, 0, at.Location())
		if int(cand.Weekday()) != s.DayOfWeek {
			continue
		}
		if cand.After(at) {
			continue
		}
		return cand, cand.Add(time.Duration(s.DurationMinutes) * time.Minute), true
	}
	return time.Time{}, time.Time{}, false
}

// IsActiveNow returns true when sabbath is in effect right now per
// either the manual ActiveOn flag (date matches today) or the
// schedule window (now ∈ [start, end)). Used by every server-side
// gate.
//
// Errors are treated as "not active" — failing-open is safer than
// failing-closed (a missing sidecar shouldn't suppress the user's
// expected notifications).
func IsActiveNow(vaultRoot string) bool {
	s, err := Load(vaultRoot)
	if err != nil {
		return false
	}
	return s.IsActiveAt(time.Now())
}

// IsActiveAt is the pure function behind IsActiveNow — useful for
// tests and for callers that already have a State and a specific
// instant in mind.
func (s State) IsActiveAt(at time.Time) bool {
	if s.ActiveOn != "" && s.ActiveOn == at.Format("2006-01-02") {
		return true
	}
	start, end, ok := scheduleWindow(s.Schedule, at)
	if !ok {
		return false
	}
	return !at.Before(start) && at.Before(end)
}

// RemainingMinutes returns how many minutes are left in the current
// sabbath window. For manual-only activations (ActiveOn matches
// today, no schedule window) it falls back to "until midnight" — the
// behavior the old client assumed. Returns 0 when not active.
func RemainingMinutes(s State, at time.Time) int {
	if !s.IsActiveAt(at) {
		return 0
	}
	if start, end, ok := scheduleWindow(s.Schedule, at); ok && !at.Before(start) && at.Before(end) {
		return int(end.Sub(at) / time.Minute)
	}
	// Manual-only path: midnight tomorrow.
	tomorrow := time.Date(at.Year(), at.Month(), at.Day()+1, 0, 0, 0, 0, at.Location())
	return int(tomorrow.Sub(at) / time.Minute)
}

// NormalizeSchedule clamps a Schedule's fields to valid ranges so a
// malformed PUT can't poison the on-disk state. Duration <=0 falls
// back to 1440 (24h) so a client sending an unset value still gets
// the default window.
func NormalizeSchedule(s Schedule) Schedule {
	if s.DayOfWeek < 0 || s.DayOfWeek > 6 {
		s.DayOfWeek = 0
	}
	if s.StartHour < 0 || s.StartHour > 23 {
		s.StartHour = 0
	}
	if s.StartMinute < 0 || s.StartMinute > 59 {
		s.StartMinute = 0
	}
	if s.DurationMinutes <= 0 {
		s.DurationMinutes = 1440
	}
	if s.DurationMinutes > 7*1440 {
		s.DurationMinutes = 7 * 1440 // cap at a week
	}
	return s
}

// IsActiveToday is kept as a thin alias for IsActiveNow. The old
// name suggested calendar-day semantics which the schedule windows
// break; new code should call IsActiveNow.
//
// Deprecated: use IsActiveNow.
func IsActiveToday(vaultRoot string) bool {
	return IsActiveNow(vaultRoot)
}
